Prg 118 backfill provider name field on be side

chodorowicz opened the pull request:
Closes https://linear.app/metabase/issue/PRG-118/backfill-provider-name-field-on-be-side

### Description

- Create `provider-detection` module on the BE side
- Add detection step to the startup sequence (health check step)
- Send provider config(names + regex) through the API to FE. 
- Use this config ☝️  instead of one hardcoded  on the FE side

### How to verify

You can repeat the tests outlined in the video 👇 

### Demo

https://www.loom.com/share/4136206e91d7464c9fd1c74eb22727f4

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

chodorowicz committed:
Populate provider_name field on startup health check

chodorowicz committed:
Serve provider patterns from BE side

chodorowicz committed:
cleanup tests and expand clojure host provider detection

chodorowicz committed:
Fix provider detection test

dpsutton commented on src/metabase/warehouses/provider_detection.clj:
```diff
@@ -0,0 +1,54 @@
+   {:name "Render" :pattern #"\.render\.com$"}
+   {:name "Scaleway" :pattern #"\.scw\.cloud$"}
+   {:name "Supabase" :pattern #"pooler\.supabase\.com|\.supabase\.co$"}
+   {:name "Timescale" :pattern #"\.tsdb\.cloud|\.timescale\.com$"}])
```
Does the binding of `$` in the Timescale regex apply to both patterns joined by the `|` or just the latter one?

dpsutton commented on src/metabase/driver/common.clj:
```diff
@@ -394,3 +395,8 @@
+
+;; Driver-specific extra-info methods
+
+(defmethod driver/extra-info :postgres [_]
```
It's strange that this exports all providers as postgres providers but those aren't marked as such. That list seems to invite additions but there's no mechanism for which ones are postgres, which ones mongo, etc. And I don't see how third party authors could extend it. That's an acceptable limitation for the time being i suppose, but seems strange that everything is essentially hardcoded to postgres.

dpsutton commented on src/metabase/warehouses/models/database.clj:
```diff
@@ -212,6 +213,26 @@
+                                           (:name database) (:id database)
+                                           (:provider_name database) detected-provider))
+                 ;; Use direct query to avoid triggering search index updates
+                 (t2/query {:update :metabase_database
```
this should just be `(t2/update! :model/Database (:id database) {:provider_name "whatever})`

dpsutton replied:
that error is unrelated

dpsutton reviewed (dismissed):
some significant changes requested. Foremost is that we are far too chatty and naive on startup about always detecting, always updating, and always logging regardless if we are doing anything new.

iethree commented on frontend/src/metabase/databases/components/DatabaseHostnameWithProviderField/provider-config.mock.ts:
if our tests are complex enough to need mocks like this we usually make a test/ subdirectory

iethree reviewed (approved):
✨ Frontend looks good to me!

chodorowicz committed:
Fix regexes for supabase and timescale

chodorowicz committed:
adjust module api exports to standards

chodorowicz committed:
mark private function ins provider-detection namespace

andsalves reviewed (approved):
Looks good to me!

chodorowicz committed:
extend provider values to null and empty, skip backfilling on empty provider_name

chodorowicz committed:
Don't hardcode postgres in health check provider detection

chodorowicz committed:
Make providers api dynamic and ready to support more engines

chodorowicz committed:
Sort lines providers and engines

chodorowicz committed:
Move hostname field tests to a test folder

chodorowicz committed:
memoize detected provider

chodorowicz committed:
Fix for tests for private functions

iethree commented on frontend/src/metabase-types/api/settings.ts:
```diff
@@ -52,6 +52,29 @@ export const engineKeys = [
+export const providerNames = [
+  "Aiven",
+  "Amazon RDS",
+  "Supabase",
```
fun fact: I worked at blockbuster video for a summer and I got REALLY good at alphabetizing things 😁  🧓

chodorowicz committed:
Fix import of DatabaseProviderName type

dpsutton commented on src/metabase/warehouses/models/database.clj:
```diff
@@ -212,6 +213,26 @@
+                 (t2/query {:update :metabase_database
+                            :set {:provider_name detected-provider}
+                            :where [:= :id (:id database)]}))
+               (catch Throwable provider-e
```
oh i must have missed that.

chodorowicz committed:
Use update function for setting provider_name on health check

chodorowicz committed:
Fix typo in test case

chodorowicz committed:
Remove duplicate entry

chodorowicz commented:
@cubic-dev-ai review this PR

chodorowicz committed:
simplify detect-provider-from-database call

chodorowicz committed:
Move postgres extra-info to postgres driver module

chodorowicz committed:
Simplify provider update check on startup, use null value for unidentified

chodorowicz committed:
match provider also using engine

chodorowicz commented on src/metabase/warehouses/models/database.clj:
```diff
@@ -212,6 +213,23 @@
            (analytics/inc! :metabase-database/status {:driver engine :healthy true})
 
+           ;; Detect and update provider name if not already set (only for databases that support provider detection)
+           (when (and (get (driver/extra-info driver) :providers)
```
I'm checking also for engine within `provider-detection` module https://github.com/metabase/metabase/pull/62420/commits/3c2bdf5e64ca777ee31a487214b72c8df07bc11e

dpsutton committed:
bit of cleanup

- just move the regexes into the postgres driver namespace
- remove regexes and turn them into strings since most databases will put
  them in yaml
- just test the public api
- simplify all the host and details checks. Database is a very standard
  shape on the backend

chodorowicz committed:
allowing setting null to provider match is null, cleanup unused imports

dpsutton committed:
move postgres tests to postgres driver; add a few tests for yaml

# Conflicts:
#	test/metabase/driver/postgres_test.clj
